Mathlib/Tactic/NormNum/NatFactorial.lean

```diff
@@ -0,0 +1,138 @@
+/-
+Copyright (c) 2023 Sebastian Zimmer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Zimmer
+-/
+import Mathlib.Data.Nat.Factorial.Basic
+import Mathlib.Tactic.NormNum
+
+/-! # `norm_num` extensions for factorials
+
+Extensions for `norm_num` that compute `Nat.factorial`, `Nat.ascFactorial` and `Nat.descFactorial`.
+
+This is done by reducing each of these to `ascFactorial`, which is computed using a divide and
+conquer strategy that improves performance and avoids exceeding the recursion depth.
+
+-/
+
+namespace Mathlib.Meta.NormNum
+
+open Nat Qq Lean Elab.Tactic Qq Meta
+
+lemma asc_factorial_aux (n l m a b : ℕ) (h₁ : n.ascFactorial l = a)
+    (h₂ : (n + l).ascFactorial m = b) : n.ascFactorial (l + m) = a * b := by
+  rw [← h₁, ← h₂]
+  symm
+  apply ascFactorial_mul_ascFactorial
+
+/-- Calculate `n.ascFactorial l` and return this value along with a proof of the result. -/
+partial def proveAscFactorial (n l : ℕ) (en el : Q(ℕ)) :
+    (result : ℕ) × (eresult : Q(ℕ)) × Q(($en).ascFactorial $el = $eresult) :=
+  if l ≤ 50 then
+    have res : ℕ := n.ascFactorial l
+    have eres : Q(ℕ) := mkRawNatLit (n.ascFactorial l)
+    have : ($en).ascFactorial $el =Q $eres := ⟨⟩
+    ⟨res, eres, q(Eq.refl $eres)⟩
+  else
```

Comment:
I'm pretty sure your test-case doesn't cover this branch; can you add some tests with `l >= 50`?